### DISCUSSION TITLE
How to stop propagation of touchmove events

### DISCUSSION THREAD

**Discussion by esm7** [Leaflet/Leaflet#8972] [General]
Hi all,

I am the author of Obsidian Map View (https://github.com/esm7/obsidian-map-view), a plugin for Obsidian.md that utilizes Leaflet.js for visualizing notes with geolocations.
To make the plugin work with the mobile version of Obsidian back in the old days of Leaflet 1.7.x, I used the following code:
```
        this.display.viewDiv.addEventListener('touchmove', (ev) => {
            ev.stopPropagation();
        });
```

`viewDiv` is the div that hosts the Leaflet.js map instance.

This used to take care of not propagating touch events upwards to the Obsidian interface when dragging the map, so dragging the map would not activate various Obsidian gestures.
However, this hack stopped working with the release of Leaflet 1.8, and I just stuck with 1.7.x for a while.
I think it's time to upgrade, but I was unable to find any hints on the docs or Github issues about why this stopped working and what I should do instead.
With this code in Leaflet 1.8 and above, the map does not drag at all. And if I remove this `stopPropagation`, I am back to the original problem of various gestures of the host app (Obsidian) firing when I pan the map.

I tried to track using the dev tools what events are being handled, and as another attempt, I tried instead to catch and stop `touchstart` and `touchend`. This yielded better results (the map drags without Obsidian interrupting), but causes various other problems, like pinch zoom not working properly.

Any tips on how to properly accomplish what I'm trying to do -- allow panning the map without triggering parent elements events -- would be welcome!

**Comment by IvanSanchez** [comment]
After having had a look at the changelog, I think it might be https://github.com/Leaflet/Leaflet/pull/7059 - Leaflet 1.8 prefers handling `pointer` events instead of `touch` events. So you should try to stop listening to (and stopping propagation of) `touch*` events, and change that to handle `pointer*` events instead.

**Reply by esm7** [reply]
Thanks for checking this out!
I can't figure it out though.
Switching the original `stopPropagation` of `touchmove` to `pointerdown`, `pointerup` and `pointermove` does not seem to make a difference: the Leaflet map pans fine, but it triggers various behaviors of the host app.
I might be missing something big here, but it seems to me like the question is not what events to block -- as this depends on what the host app (Obsidian) listens to, and not what Leaflet listens to -- but rather why blocking `touchmove` upwards makes Leaflet stop working starting 1.8.

**Comment by esm7** [comment]
Hi, I'm trying to take another shot at upgrading Obsidian Map View to the latest Leaflet.js, but not yet sure how to solve this problem.
- The old code of using `ev.stopPropagation()` for `touchmove`, that worked in the old Leaflet 1.7.x, causes inability to drag the map on mobile after upgrading to later Leaflet versions.
- Removing this code makes it impossible to use the map because the host apps handles the touches in ways that interfere (e.g. opening drawers from the left and right).

To my understanding, the fact `stopPropagation` causes the map to stop panning probably means that Leaflet now handles the events somewhere up the path and not within the map element; any clue how I can make the Leaflet handler still work while disabling the host app event handling?

**Reply by esm7** [reply]
I'll also add that I tried walking through event handlers at various levels, and print the path of the `touchmove` event, and I can't see any indication that Leaflet is handling it above the map div level. Which means I'm clueless about why `ev.stopPropagation()` is not allowing the map to be dragged since Leaflet 1.8 :(

**Reply by Falke-Design** [reply]
Hi @esm7 can you please create some fiddle, it is very hard to help without some running code. http://www.leafletjs.com/edit.html

**Reply by esm7** [reply]
I'd love to provide any information that can help me solve the problem, but what kind of fiddle can I create for this problem? The issue is that my Leaflet.js instance runs inside an Obsidian.md plugin, I have no control over the host app, what can I provide to help you guys simulate the relevant issue? :(

**Reply by Falke-Design** [reply]
@esm7 please send me per mail ([design.falke@gmail.com](mailto:design.falke@gmail.com)) a detailed "tutorial" how I can install Obsidian and how to integrate the plugin while working on it and with some sample data.

I will try to take a look in the next 2 weeks.

**Reply by esm7** [reply]
Thank you, this would be amazing! Sending now :pray:

**Reply by Falke-Design** [reply]
@esm7 it is an easy fix. Just replace `this.display.viewDiv` with `document`:

```
document.addEventListener('touchmove', (ev) => {
    ev.stopPropagation();
});
```

**Reply by esm7** [reply]
This works! I'm so thankful :pray: 
For my own education, I'd love to understand why this works and what changed in Leaflet 1.8 that required this change!
Generally, maybe it's my own JS ignorance, but I'm confused about the usage of `stopPropagation` in the `document` level: isn't `document` the top level, and as such, an event that bubbles upwards has nowhere more to propagate anyway? There's obviously something important I'm missing here :smile:

**Reply by Falke-Design** [reply]
> document the top level, and as such, an event that bubbles upwards

exactly. Leaflet listens on the map-DOM element for touchevents and you stopped propagation and executing of the event. Now with `document` Leaflet still gets the touchevents on the map-DOM element and then it bubbles upwards. Because of the change now document doesn't execute the listeners of Obsidian anymore but the elements below still can listen on the event.

**Reply by esm7** [reply]
For future reference, in case anyone else ever needs a similar solution and finds this discussion, I also added a check that the target of the event is indeed the map div. Without doing so, the above snippet completely disables Obsidian's touch gestures, regardless if the user is interacting with the map or is currently using the plugin at all.
This seems to work:
```
        document.addEventListener('touchmove', (ev: TouchEvent) => {
			const mapDiv = this.display.mapDiv;
			const targetNode = ev.target as Node;
			const isEventOnMap = ev.target === mapDiv || mapDiv.contains(targetNode);
			if (isEventOnMap)
				ev.stopPropagation();
        });
```
